go/consensus/tendermint/apps/roothash/liveness.go

```diff
@@ -0,0 +1,94 @@
+package roothash
+
+import (
+	"fmt"
+
+	beacon "github.com/oasisprotocol/oasis-core/go/beacon/api"
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/signature"
+	tmapi "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/api"
+	registryState "github.com/oasisprotocol/oasis-core/go/consensus/tendermint/apps/registry/state"
+	registry "github.com/oasisprotocol/oasis-core/go/registry/api"
+	roothash "github.com/oasisprotocol/oasis-core/go/roothash/api"
+	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+)
+
+// processLivenessStatistics checks the liveness statistics for the last epoch and penalizes any
+// nodes that didn't satisfy the liveness condition.
+func processLivenessStatistics(ctx *tmapi.Context, epoch beacon.EpochTime, rtState *roothash.RuntimeState) error {
+	if rtState.ExecutorPool == nil || rtState.LivenessStatistics == nil || rtState.Suspended {
+		return nil
+	}
+
+	// Skip evaluation if the number of total live rounds is below the set minimum.
```

Comment:
In order to catch pathological mis-configuration, this should either special case check for `TotalRounds == 0` or enforce `MinLiveRoundsForEvaluation >= 1`

Reply:
Agreed, fixed.